fission-core/library/Fission/App/Name/Types.hs

```diff
@@ -0,0 +1,323 @@
+-- | The name of a Fission app (i.e. the subdomain in Fission's app namespace)
+module Fission.App.Name.Types
+  ( Name
+  , raw
+  , mkName
+  ) where
+
+import qualified RIO.ByteString.Lazy    as Lazy
+import qualified RIO.Text               as Text
+
+import           Data.Swagger           hiding (name)
+import           Servant.API
+
+import           Fission.Prelude
+
+import qualified Fission.App.Name.Error as App.Name
+import           Fission.URL.Validation
+
+newtype Name = Name { raw :: Text }
+  deriving newtype (Eq, Show, Display)
+
+mkName :: Text -> Either App.Name.Invalid Name
+mkName txt =
+  if isValid normalized
+    then Right $ Name normalized
+    else Left App.Name.Invalid
+
+  where
+    normalized = Text.toLower txt
```

Comment:
@walkah here we are exporting the _type_ but not the _data constructor_. This means that to create a `Name` outside of this module, you _must_ go through `mkName`. This is a simple technique for enforcing runtime validation called a "smart constructor".

Reply:
Similar to `mkUsername` / Username right?

Reply:
Yup the exact same technique